packages/arb-avm-cpp/data_storage/src/messageentry.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <data_storage/messageentry.hpp>
+
+#include "value/referencecount.hpp"
+#include "value/utils.hpp"
+
+MessageEntry extractMessageEntry(const rocksdb::Slice key,
+                                 const rocksdb::Slice value) {
+    // Extract key
+    const char* keyptr = key.data();
+    auto sequence_number = deserializeUint256t(keyptr);
+
+    // Extract message entry
+    auto entry_vector = std::vector<const unsigned char>{
+        value.data(), value.data() + value.size()};
+
+    return deserializeMessageEntry(sequence_number, entry_vector);
+}
+
+MessageEntry deserializeMessageEntry(
+    const uint256_t sequence_number,
+    std::vector<const unsigned char> entry_vector) {
+    auto current_iter = entry_vector.begin();
+
+    auto inbox_hash = extractUint256(current_iter);
+    auto block_height = extractUint64(current_iter);
+    auto last_message_in_block = current_iter[0] == 1;
+    auto message_size = extractUint64(current_iter);
+    uint64_t remaining_size = entry_vector.end() - current_iter;
+    if (remaining_size < message_size) {
+        message_size = remaining_size;
+    }
+    auto message = rocksdb::Slice{reinterpret_cast<const char*>(*current_iter),
```

Comment:
Unless I'm missing something, you're creating a dangling pointer here since `entry_vector` is passed in by value, and `rocksdb::Slice` is non-owning. We likely should just keep message as a vector since I'm assuming we actually want ownership semantics here